src/fifolock.jl

```diff
@@ -0,0 +1,143 @@
+@static if VERSION >= v"1.10"
+
+const LOCKED_BIT = 0b01
+
+"""
+    FIFOLock()
+
+Like Base.ReentrantLock, but with strict FIFO ordering.
+"""
+mutable struct FIFOLock <: AbstractLock
+    @atomic locked_by::Union{Task, Nothing}
+    reentrancy_cnt::UInt32
+    @atomic havelock::UInt8
```

Comment:
Why doesn't `reentrancy_cnt` need to be `@atomic`?

I see that it's always guarded by having locked `cond_wait` (I think), but so is `locked_by` i think?

Reply:
yeah but those can be two different tasks over time, meaning this can be written and read from different Threads, right? Doesn't that need to be `@atomic` also, since our `atomic` keyword implies `volatile`?

Reply:
There are (probably a good few) memory fences between two different tasks accessing this field, so I believe it is safe. May be worth noting that `Base.ReentrantLock` is very similar in this regard.

Reply:
thanks, i forgot that the lock/unlock on the condition variable's lock introduces a fence as well, so it's okay. thanks